### PR TITLE
[XLA:CPU][oneDNN] Convolution XLA HLO Pattern Matcher with oneDNN custom call rewrite

### PR DESCRIPTION
**Pull request by akhilgoe**
This PR enables oneDNN library call for the matched XLA HLO Convolution pattern through custom_call instruction. In particular, this PR:

1. Adds oneDNN convolution rewriter pass that will rewrite HLO Convolution to oneDNN Convolution.
2. Refactors backend config to enhance code reusability.
3. Adds a convolution test file to verify rewrite and execution result

### CODE REVIEW

**Comment by tdanyluk** [comment]
I think that I'll wait for @penpornk's review.

**Comment by kamaljeeti** [comment]
Hi @penpornk , can you look into this once? Thanks.

**Comment by kamaljeeti** [comment]
Hi @penpornk , any update on this PR? Thanks.

**Comment by kamaljeeti** [comment]
Hi @jvstokes, any update on this PR? Thanks.

File: xla/service/cpu/backend_config.proto
Comment (penpornk) [inline xla/service/cpu/backend_config.proto:18]: Note that we usually can't delete fields in protobuf (can only add new fields and, at most, mark old fields as deprecated -- to maintain forward/backward compatibility). But since these protos are only used in internal XLA C++ files, it should be okay to drop the compatibility here.

File: xla/service/cpu/backend_config.proto
Comment (penpornk) [inline xla/service/cpu/backend_config.proto:46]: Please prefix these names with `OneDnn` as well.

File: xla/service/cpu/backend_config.proto
Comment (penpornk) [inline xla/service/cpu/backend_config.proto:1]: Please refactor oneDNN-specific proto definitions into a new file, e.g., onednn_config.proto.

File: xla/service/cpu/onednn_convolution_rewriter.cc
Comment (penpornk) [inline xla/service/cpu/onednn_convolution_rewriter.cc:1]: Nit: Please add a blank line below this line.

File: xla/service/cpu/onednn_convolution_rewriter.h
Comment (penpornk) [inline xla/service/cpu/onednn_convolution_rewriter.h:1]: Nit: Please add a blank line below this line.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and I'm very sorry for the delay!

Feedback for future PRs like this:
It would have been better if this PR were split into two PRs
1. Protobuf refactoring
1. Adding convolution support.

File: xla/service/algebraic_simplifier.cc
Comment (akhilgoe) [reply xla/service/algebraic_simplifier.cc:9014]: Yes, the FloatSupport handles the conversions. However, because Algebraic Simplifier precedes the oneDNN rewriter, there can be instances where the Algebraic Simplifier may transform oneDNN rewritable BF16 convolutions into BF16 Muls and Dots which can create problems.
This can be removed once BF16 Muls and Dots are fully supported.

File: xla/service/cpu/onednn_convolution_rewriter.h
Comment (akhilgoe) [reply xla/service/cpu/onednn_convolution_rewriter.h:1]: Done

File: xla/service/cpu/onednn_convolution_rewriter.cc
Comment (akhilgoe) [reply xla/service/cpu/onednn_convolution_rewriter.cc:1]: Done

File: xla/service/cpu/backend_config.proto
Comment (akhilgoe) [reply xla/service/cpu/backend_config.proto:1]: Done

File: xla/service/cpu/backend_config.proto
Comment (akhilgoe) [reply xla/service/cpu/backend_config.proto]: Done

File: xla/service/algebraic_simplifier.cc
Comment (akhilgoe) [reply xla/service/algebraic_simplifier.cc]: Yes, you are right! Thank you for pointing out. I have made the changes.

**Comment by akhilgoe** [comment]
Hi @penpornk, Thank you for the review! I think I have addressed all your comments and resolved conflicts. Can you please take a look again? Thanks!

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the changes and sorry for the delay! I was on vacation. I have one last comment. :)

File: xla/service/algebraic_simplifier.cc
Comment (akhilgoe) [reply xla/service/algebraic_simplifier.cc:9016]: Yes, I think that would be much better! Thanks for the suggestion.

**Comment by akhilgoe** [comment]
Thanks for the suggestions @penpornk . I have refactored the oneDNN re-writability check. Please let me know your thoughts.

**Review by penpornk — APPROVED** [review]
Thank you for the changes!

**Review by penpornk — APPROVED** [review]

**Comment by sergeykozub** [comment]
This fails internal presubmits, even though CI passes on GitHub - specifically, TensorFloat32Test.test_conv2d_backprop_filter - and I have hard time reproducing it locally, will continue looking.
Note: I also had to add "package xla.cpu;" to "onednn_config.proto" to be able to compile it.

**Comment by akhilgoe** [comment]
Thanks @sergeykozub for the info! Can you please point me to the test? I'll try to reproduce it too.

**Comment by akhilgoe** [comment]
Hi @sergeykozub and @penpornk,

I reviewed the failed TensorFlow test. It compares F32 output against F64 output and the 2 outputs are expected to match with a tolerance of 1e-5. However, oneDNN algorithms may form large accumulation chains for some inputs to utilize the hardware efficiently, potentially causing some precision loss. Therefore, we need to adjust the test tolerance accordingly. I have created this TensorFlow PR to make the necessary adjustments. Could you please review it?

https://github.com/tensorflow/tensorflow/pull/68745

Thanks you!

**Review by akuegel — APPROVED** [review]
Approving again, now that the tensorflow test required precision has been adjusted.

**Comment by akhilgoe** [comment]
Hi @penpornk and @akuegel, just wondering if the PR would need a re-approval to get merged. Thanks!

File: xla/service/cpu/onednn_config.proto
Comment (akuegel) [inline xla/service/cpu/onednn_config.proto:63]: Why not 2, 3, 4 here?

File: xla/service/cpu/onednn_config.proto
Comment (akuegel) [inline xla/service/cpu/onednn_config.proto:78]: 4 instead of 8?

File: xla/service/cpu/onednn_config.proto
Comment (akuegel) [inline xla/service/cpu/onednn_config.proto:79]: Given there are so few ids, I think the comment for next id is not needed

File: xla/service/cpu/onednn_config.proto
Comment (akhilgoe) [reply xla/service/cpu/onednn_config.proto:78]: Currently, in the main branch OneDnnMatmulConfig already has ids from 1 through 7 in use, and in the commits to proto files in the repo I see authors don't re-use a previously used id so I thought to give new items new and unique ids.

Comment (akuegel) [reply xla/service/cpu/onednn_config.proto:78]: This is a new message, so new ordering can be used. Only within one message, different ids need to be used. I am not aware of people actively avoiding to reuse ids from different messages.

Comment (akhilgoe) [reply xla/service/cpu/onednn_config.proto:78]: Here, we refactor a lot of stuff out from OneDnnMatmulConfig. I think because in this PR we moved it to a new file it shows up as a new message, but we do have this message in the main branch. This is what it looks like in main right now:

```
message OneDnnMatMulConfig {
  bool transpose_a = 1;
  bool transpose_b = 2;
  // These enum needs to be mapped to oneDNN enum for post_op algorithm.
  // TODO(intel-tf): Add kinds supported by oneDNN.
  enum FusionKind {
    UNDEFINED = 0;
    BIAS = 1;
    RELU = 2;
    TANH = 3;
    GELU_ERF = 4;
    GELU_TANH = 5;
    BINARY_ADD = 6;
    LINEAR = 7;
    ELU = 8;
    RELU6 = 9;
    SIGMOID = 10;
  }
  repeated FusionKind fused_ops = 3;
  bool bias_broadcast = 4;
  // To avoid protobuf failures for specific decimal values,
  // the original float value alpha is type-casted to int32.
  int32 alpha_typecast = 5;
  bool weights_prepacked = 6;
  bool user_scratchpad = 7;
}
```

Comment (akuegel) [reply xla/service/cpu/onednn_config.proto:78]: Ok, I understand now that this is just a proto message that got moved from backend_config.proto to onednn_config.proto. In that case, I would suggest adding "reserved" tags for those fields which will not exist anymore after the move.
e.g.
```
reserved 4;  // Was bias_broadcast
```

**Comment by akhilgoe** [comment]
Hi @akuegel, I have made the changes. Please take a look. Thanks!

**Review by akuegel — APPROVED** [review]

**Pull request closed without merging**
